### PR TITLE
.github/workflows/test.yml: Improve test workflow

### PR DESCRIPTION
**Pull request by EduKav1813**

### CODE REVIEW

**Review by m-iwanicki — CHANGES_REQUESTED** [review]
Why did you change indentation to 4 spaces? we have 2 spaces everywhere else.

**Comment by EduKav1813** [comment]
@m-iwanicki Changed to 2 spaces.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -51,15 +42,48 @@ jobs:
...
           ./qemu-run.sh nographic os &
-      - name: Create directory for logs
-        shell: bash
-        id: log_dirs
-        run: |
-          timestamp=$(date -u +%Y-%m-%dT%H:%M:%S%Z)
-          directory="/tmp/dts-test-ci-${timestamp}"
-          mkdir $directory
-          echo "directory=$directory" >> "$GITHUB_OUTPUT"
-      - name: Run tests
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:62]: Why did you remove this step? You should probably create folder per each robot run e.g.:

```
mkdir -p "/tmp/dst-test-ci-${{ github.run_id }}/${{ matrix.include.test }}"
```

so we can temporarily store logs in case of errors (otherwise cleanup job would remove logs)

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -76,41 +100,13 @@ jobs:
...
-          -v dpp_logs_key:$MSI_LOGS -v netboot_utilities_support:True \
-          -v dts_ipxe_link:http://${ip_addr}:4321/dts.ipxe \
-          -i "msi" dts/dts-e2e.robot 2>&1 | tee $LOG_DIR/output_msi.log | grep "| PASS |\|| FAIL |"
-
-          robot -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v snipeit:no \
-          -v dpp_password:$DPP_PASSWORD -v dpp_download_key:$MSI_HEADS_DOWNLOAD \
-          -v dpp_logs_key:$MSI_HEADS_LOGS -v netboot_utilities_support:True \
+          ./scripts/run.sh -L TRACE -v config:qemu -v rte_ip:127.0.0.1 -v snipeit:no \
+          -v dpp_password:$DPP_PASSWORD -v dpp_download_key:${{ matrix.config.dpp_download_key }} \
+          -v dpp_logs_key:${{ matrix.config.dpp_logs_key }} -v netboot_utilities_support:True \
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:105]: shouldn't it be `${{ matrix.include. }}` here and everywhere else?

```yml
    strategy:
      matrix:
        include:
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -3,14 +3,15 @@ name: Run DTS tests
...
+      - "main"
 jobs:
   build-dts:
     uses: ./.github/workflows/build.yml
     with:
       cacheless: false
-  run-tests:
-    name: Run DTS tests
+
+  prepare-for-tests:
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:13]: Each job is run in different session so most likely even detached processes would be killed between jobs.
You need to start ipxe server and qemu in the same job you are using it. As matrix creates multiple jobs you probably need to start qemu/ipxe before each `robot` step

**Review by m-iwanicki — CHANGES_REQUESTED** [review]
You should probably rebase on develop as there were couple changes

File: .github/workflows/test.yml (hunk 3 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: Moved here: 051caa1d67b115128723802a638aeb2cc97d31a9

File: .github/workflows/test.yml (hunk 1 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: Added the step here: 2f8993cab0d0c4e27cf009abaaf0e2eb52333868

Comment (m-iwanicki) [reply .github/workflows/test.yml]:
```yaml
      - name: Create directory for logs
        shell: bash
        id: create_log_dirs
        run: |
          export LOG_DIR="/tmp/dst-test-ci-${{ github.run_id }}/${{ matrix.include.test }}"
          mkdir -p $LOG_DIR
          echo "LOG_DIR=$LOG_DIR" >> $GITHUB_ENV
      - name: Run ${{ matrix.include.test }}
        shell: bash
        env:
          LOG_DIR: ${{ steps.log_dirs.outputs.directory }}
```

In `Run` step you are trying to access `${{ steps.log_dirs.outputs.directory }}` but previous step has id `create_log_dirs` and outputted variable name is `LOG_DIR`

Comment (m-iwanicki) [reply .github/workflows/test.yml]: Didn't check rest of steps, check and fix in all places if needed

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -3,45 +3,84 @@ name: Run DTS tests
...
-          echo -e "\n
-          #!ipxe\n
-          imgfetch --name file_kernel bzImage\n
-          imgfetch --name file_initrd dts-base-image-genericx86-64.cpio.gz\n
-          kernel file_kernel root=/dev/nfs initrd=file_initrd\n
-          boot" > ipxe/dts.ipxe
-          cd ipxe && python3 -m http.server 4321 &
+        run: ./open-source-firmware-validation/scripts/ci/ipxe-run.sh \
+          build/tmp/deploy/images/genericx86-64/dts-base-image-genericx86-64.cpio.gz \
+          build/tmp/deploy/images/genericx86-64/bzImage
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:74]:
```suggestion
          build/tmp/deploy/images/genericx86-64/bzImage &
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -3,45 +3,84 @@ name: Run DTS tests
...
+
       - name: Install requirements
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:76]: You can do this only once as we are using the same folder between jobs (unfortunately). I'm not sure if it will work if `venv` already exists

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -3,45 +3,84 @@ name: Run DTS tests
...
+
     steps:
       - name: Checkout OSFV repo
         uses: actions/checkout@v4
         with:
-          repository: 'Dasharo/open-source-firmware-validation'
-          path: 'open-source-firmware-validation'
-          submodules: 'recursive'
-          ref: 'develop'
+          repository: "Dasharo/open-source-firmware-validation"
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:65]: You can checkout this repo only once in earlier job as I'm not sure if it'll succeed if folder already exist.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -50,15 +89,16 @@ jobs:
           mkdir qemu-data
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:89]:
```suggestion
          mkdir -p qemu-data
```

Currently this line would fail if folder already exists.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -50,15 +89,16 @@ jobs:
...
         shell: bash
-        id: log_dirs
+        id: create_log_dirs
         run: |
-          timestamp=$(date -u +%Y-%m-%dT%H:%M:%S%Z)
-          directory="/tmp/dts-test-ci-${timestamp}"
-          mkdir $directory
-          echo "directory=$directory" >> "$GITHUB_OUTPUT"
-      - name: Run tests
+          export LOG_DIR="/tmp/dst-test-ci-${{ github.run_id }}/${{ matrix.include.test }}"
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:97]:
```suggestion
          LOG_DIR="/tmp/dst-test-ci-${{ github.run_id }}/${{ matrix.include.test }}"
```

`export` won't work as those command are run as a script (workflow creates temporary file with those commands and passes it to `sh`/`bash`

**Review by m-iwanicki — CHANGES_REQUESTED** [review]
next batch

File: .github/workflows/test.yml (hunk 4 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: d900bd3e09066b47e243c9db41048752713b9a05

File: .github/workflows/test.yml (hunk 5 above)
Comment (EduKav1813) [reply .github/workflows/test.yml:76]: I figured out a rather simple solution: 447c744eda1afb6755bff1aaccfeda07d4dfdb03

Also tested it locally with act

File: .github/workflows/test.yml (hunk 7 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: 0fb47d0128d6b1601e1d9bc018b68cf76406519f

File: .github/workflows/test.yml (hunk 8 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: 8ae4197eb4fb062a504f52ada2b282b47c06b852

File: .github/workflows/test.yml (hunk 6 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: Fixed: 44075c2eec038732548a90254a3a435ba1be8173

File: .github/workflows/test.yml (hunk 1 above)
Comment (EduKav1813) [reply .github/workflows/test.yml]: Fixed all LOG_DIR occurences:  60cee9b8bd69cb88a9b635884facbb6ce0687b7a

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -3,57 +3,108 @@ name: Run DTS tests
...
         shell: bash
-        id: log_dirs
+        id: create_log_dirs
         run: |
-          timestamp=$(date -u +%Y-%m-%dT%H:%M:%S%Z)
-          directory="/tmp/dts-test-ci-${timestamp}"
-          mkdir $directory
-          echo "directory=$directory" >> "$GITHUB_OUTPUT"
-      - name: Run tests
+          LOG_DIR="/tmp/dst-test-ci-${{ github.run_id }}/${{ matrix.test }}"
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:102]:
```suggestion
          LOG_DIR="/tmp/dts-test-ci-${{ github.run_id }}/${{ matrix.test }}"
```

**Review by m-iwanicki — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -3,57 +3,122 @@ name: Run DTS tests
...
+            dpp_download_key: $OPTIPLEX_DOWNLOAD
+            dpp_logs_key: $OPTIPLEX_LOGS
+            log_file: output_nc_heads.log
+
+          - test: pcengines_dpp
+            dpp_download_key: $PCENGINES_DOWNLOAD
+            dpp_logs_key: $PCENGINES_LOGS
+            log_file: output_pcengines.log
+
+          - test: pcengines_seabios
```
Comment (m-iwanicki) [inline .github/workflows/test.yml:72]: There is no separate `pcengines_seabios` tag

```
Suite 'Dts-E2E' contains no tests matching tag 'pcengines seabios'.
```

**Review by m-iwanicki — CHANGES_REQUESTED** [review]
Optiplex and pcengines_dpp failures look like problems with tests not workflow so I guess only one change is needed.

File: .github/workflows/test.yml (hunk 10 above)
Comment (EduKav1813) [reply .github/workflows/test.yml:72]: @m-iwanicki this fixed in https://github.com/Dasharo/open-source-firmware-validation/pull/636

Comment (EduKav1813) [reply .github/workflows/test.yml:72]: I also fixed some mismathed arguments in the matrix configuration

**Comment by EduKav1813** [comment]
@m-iwanicki In the last commit I made cleanup of the temporary changes. Once we will have the results of the CI runs I think we can finally merge this.

**Comment by m-iwanicki** [comment]
After run finishes **MAKE SURE** to change target branch to develop

**Review by m-iwanicki — APPROVED** [review]
Looks good, errors look like problems with tests not workflow itself. Change target branch before mergings

```
------------------------------------------------------------------------------
E2E003.005 MSI PRO Z690-A DDR-4 initial deployment (legacy -> Core... | FAIL |
No match found for 'iPXE Shell' in 30 seconds. Output:
------------------------------------------------------------------------------
E2E003.006 MSI PRO Z690-A initial deployment (legacy -> Coreboot +... | PASS |
------------------------------------------------------------------------------
E2E003.009 MSI PRO Z690-A DDR-4 update (Coreboot + UEFI -> Coreboo... | PASS |
------------------------------------------------------------------------------
E2E003.010 MSI PRO Z690-A update (Coreboot + UEFI -> Coreboot + UE... | FAIL |
No match found for 'Does it match your actual specification? (Y|n)' in 30 seconds. Output:

Gathering flash chip and chipset information...
Flash information: Test Flash Chip
Flash size: 2M
Waiting for network connection ...
Network connection have been established!
Checking if board is Dasharo compatible.
Getting platform specific GPG key... 2
Done
Waiting for system clock to be synced ...
  DPP version (coreboot + Heads) available but you don't have access
  to it, if you are interested, please visit
  https://shop.3mdeb.com/product-category/dasharo-pro-package/

Current Dasharo version: 1.0.0
Latest available Dasharo version for your subscribtion: 1.1.4
Checking Dasharo firmware checksum... Verified.
Checking Dasharo firmware
------------------------------------------------------------------------------
Dts-E2E                                                               | FAIL |
```
